apps/verticalhome/js/pin_apps.js

```diff
@@ -0,0 +1,120 @@
+'use strict';
+
+(function (exports) {
+
+  function PinAppItem(index) {
+    if (index === null || index === undefined){
+      return false;
+    }
+    this.index = index;
+    //creating pin app element
+    var pinElem = document.createElement('div');
+    pinElem.className = 'pin-app-item';
+    pinElem.setAttribute('data-index', this.index);
+    pinElem.innerHTML = "<img class='pin-app-icon'><br></img>" +
+            "<span class='title'></span>";
+    var pinList = document.getElementById('pin-apps-list');
+    var moreAppsLi = document.getElementById('moreApps');
+    pinList.insertBefore(pinElem, moreAppsLi);
+
+    this.element = pinElem;
+    this.icon = this.element.querySelector('.pin-app-icon');
+    this.title = this.element.querySelector('.title');
+  }
+
+  PinAppItem.prototype = {
+    entry: null,
+
+    bindEntry: function(entry) {
+      if (!entry){
+        return;
+      }
+      this.clear();
+      this.entry = entry;
+      this.entry.index = this.index;
+      this.targetApp = app.getAppByURL(this.entry.manifestURL);
+      this.render();
+    },
+
+    render: function() {
+      this.element.removeEventListener("click", this);
+      if(this.targetApp) {
+        this.icon.style.visibility = "visible";
+        this.title.style.visibility = "visible";
+        this.element.dataset.manifesturl = this.entry.manifestURL;
+        if(this.entry.icon) {
+          this.icon.src = this.entry.icon;
+          this.element.addEventListener("click", this);
+        }
+        if(this.entry.name) {
+          this.title.textContent = this.entry.name;
+        }
+      } else {
+          this.icon.style.visibility = "hidden";
+          this.title.style.visibility = "hidden";
```

Comment:
Indentation is wrong here. Do you ever actually anticipate having one of these things hidden because it has no target app?  Isn't that going to mess up all the navigation code?